speed.py

```diff
@@ -0,0 +1,174 @@
+
+    Parameters
+    ----------
+    record : pd.DataFrame
+        Dataframe containing the information about the record
+
+    Returns
+    -------
+    float
+        Length in seconds
+    """
+    return record.end.max() - record.start.min()
+
+
+def get_intervals_by_note(record: pd.DataFrame, duration: int = 1) -> pd.IntervalIndex:
+    """
+    Creates and returns an IntervalIndex for each time a note is played.
+
+    Parameters
+    ----------
+    record : pd.DataFrame
+        DataFrame containing the data about the notes played
+    duration : int, optional
+        Duration of the intervals, by default 1
+
+    Returns
+    -------
+    pd.IntervalIndex
+    """
+    start_times = record["start"].unique()
+
+    starts = []
+    ends = []
+    for i in start_times:
+        starts += [i]
+        ends += [i + duration]
+
+    return pd.IntervalIndex.from_arrays(starts, ends, closed="left")
+
+
+def get_intervals(record: pd.DataFrame, duration: int = 1) -> pd.IntervalIndex:
+    """
+    Creates and returns an IntervalIndex starting from the first note.
+
+    Parameters
+    ----------
+    record : pd.DataFrame
+        DataFrame containing data about the notes played
+    duration : int, optional
+        The duration of the intervals, by default 1
+
+    Returns
+    -------
+    pd.IntervalIndex
+    """
+    record_interval = pd.Interval(0, record.end.max(), closed="left")
+    intervals = pd.interval_range(
+        start=record_interval.left,
+        end=record_interval.right,
+        freq=duration,
+        closed="left",
+    )
+    return intervals
+
+
+def notes_per_interval(record: pd.DataFrame, duration: int = 1) -> pd.DataFrame:
+    """
+    Creates and returns a DataFrame containing the number of notes played in each interval.
+
+    Parameters
+    ----------
+    record : pd.DataFrame
+        DataFrame contining data about the notes played
+    duration : int, optional
+        The duration of the intervals in which we want to count the notes, by default 1
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of notes played in each interval
+    """
+    intervals = get_intervals(record, duration)
+    times = record[["start", "pitch"]].groupby("start").count()
+    nps = times.groupby(pd.cut(times.index, intervals)).sum() / duration
+    nps["count"] = nps["pitch"]
+    nps = nps.drop("pitch", axis=1)
+    return nps
+
+
+def count_notes_in_interval(record: pd.DataFrame, interval: pd.Interval) -> int:
+    """
+    Counts the number of notes played in a given interval.
+
+    Parameters
+    ----------
+    record : pd.DataFrame
+        DataFrame containing data about the notes played
+    interval : pd.Interval
+        Interval in question
+
+    Returns
+    -------
+    int
+        Number of notes played in the given interval
+    """
+    return record.loc[(record["start"] >= interval.left) & (record["end"] < interval.right)]["pitch"].count()
```

Comment:
In terms of clean code, my opinion here is that this shouldn't be a separate function . You created 15 lines of documentation for a single line of code, but the documentation only describes inputs and outputs - it doesn't tell me anything about what the code is actually doing. Here's an alternative approach:

```python
# For notes within the interval ...
ids = (record.start >= interval.left) & (record.end < interval.right)
# ... count all the pitches
pitch_count = record[ids]["pitch"].count()
```
And I would keep it like that in `count_notes_by_intervals`, even at the cost of breaking L137 into 4 or 5 lines.

Again, this is my opinion, there's no difference in logic, only in the way it reads. At the same I think that readability is the most important factor when working on complex projects, and even small improvements can make a huge difference in the long term :)

Reply:
Moved the functionality to the other function.